consensus-types/hdiff/state_diff.go

```diff
@@ -0,0 +1,2040 @@
+			WithdrawalCredentials: slices.Clone(d.WithdrawalCredentials),
+			Amount:                d.Amount,
+			Signature:             slices.Clone(d.Signature),
+			Slot:                  d.Slot,
+		}
+	}
+	return nil
+}
+
+func diffPendingPartialWithdrawals(diff *stateDiff, source, target state.ReadOnlyBeaconState) error {
+	tPendingPartialWithdrawals, err := target.PendingPartialWithdrawals()
+	if err != nil {
+		return err
+	}
+	tlen := len(tPendingPartialWithdrawals)
+	tPendingPartialWithdrawals = append(tPendingPartialWithdrawals, nil)
+	var sPendingPartialWithdrawals []*ethpb.PendingPartialWithdrawal
+	if source.Version() >= version.Electra {
+		sPendingPartialWithdrawals, err = source.PendingPartialWithdrawals()
+		if err != nil {
+			return err
+		}
+	}
+	tPendingPartialWithdrawals = append(tPendingPartialWithdrawals, sPendingPartialWithdrawals...)
+	index := kmpIndex(len(sPendingPartialWithdrawals), tPendingPartialWithdrawals, helpers.PendingPartialWithdrawalsEqual)
+	diff.pendingPartialWithdrawalsIndex = uint64(index)
+	diff.pendingPartialWithdrawalsDiff = make([]*ethpb.PendingPartialWithdrawal, tlen+index-len(sPendingPartialWithdrawals))
+	for i, d := range tPendingPartialWithdrawals[len(sPendingPartialWithdrawals)-index : tlen] {
+		diff.pendingPartialWithdrawalsDiff[i] = &ethpb.PendingPartialWithdrawal{
+			Index:             d.Index,
+			Amount:            d.Amount,
+			WithdrawableEpoch: d.WithdrawableEpoch,
+		}
+	}
+	return nil
+}
+
+func diffPendingConsolidations(diff *stateDiff, source, target state.ReadOnlyBeaconState) error {
+	tPendingConsolidations, err := target.PendingConsolidations()
+	if err != nil {
+		return err
+	}
+	tlen := len(tPendingConsolidations)
+	tPendingConsolidations = append(tPendingConsolidations, nil)
+	var sPendingConsolidations []*ethpb.PendingConsolidation
+	if source.Version() >= version.Electra {
+		sPendingConsolidations, err = source.PendingConsolidations()
+		if err != nil {
+			return err
+		}
+	}
+	tPendingConsolidations = append(tPendingConsolidations, sPendingConsolidations...)
+	index := kmpIndex(len(sPendingConsolidations), tPendingConsolidations, helpers.PendingConsolidationsEqual)
+	diff.pendingConsolidationsIndex = uint64(index)
+	diff.pendingConsolidationsDiffs = make([]*ethpb.PendingConsolidation, tlen+index-len(sPendingConsolidations))
+	for i, d := range tPendingConsolidations[len(sPendingConsolidations)-index : tlen] {
+		diff.pendingConsolidationsDiffs[i] = &ethpb.PendingConsolidation{
+			SourceIndex: d.SourceIndex,
+			TargetIndex: d.TargetIndex,
+		}
+	}
+	return nil
+}
+
+// applyValidatorDiff applies the validator diff to the source state in place.
+func applyValidatorDiff(source state.BeaconState, diff []validatorDiff) (state.BeaconState, error) {
+	sVals := source.Validators()
+	if len(sVals) < len(diff) {
+		return nil, errors.Errorf("target validators length %d is less than source %d", len(diff), len(sVals))
+	}
+	for _, d := range diff {
+		if d.index > uint32(len(sVals)) {
```

Comment:
I thought this may be `>=` but im not 100% sure

Reply:
nope, it's >, if the diff adds validators then they need to be added sequentially at the end of the source. So `d.index == len(sVals)`  is allowed and is the fist addition, which incidentally is used in the very next line